main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,20 +40,350 @@ import ca.uwaterloo.flix.util.Validation._
+        Some(MkClosureRef(ref, freeVars.map(fv => varSyms.get(fv.sym) match {
+          case None => fv
+          case Some(sym) => FreeVar(sym, fv.tpe)
+        }), tpe, loc))
+      /* Substitute inside expression */
+      case _: LoadExpression => None
+      case _: StoreExpression => None
+      case Unit => None
+      case True => None
+      case False => None
+      case Char(_) => None
+      case Float32(_) => None
+      case Float64(_) => None
+      case Int8(_) => None
+      case Int16(_) => None
+      case Int32(_) => None
+      case Int64(_) => None
+      case BigInt(_) => None
+      case Str(_) => None
+      case Ref(sym, _, _) => None
+      case Lambda(_, _, _, _) => None
+      case Hook(_, _, _) => None
+      case MkClosure(_, _, _, _) =>
+        throw InternalCompilerException(s"Unexpected expression $e after lambda lifting")
+      case ApplyRef(_, _, _, _) => None
+      case ApplyTail(_, _, _, _, _) => None
+      case ApplyHook(_, _, _, _) => None
+      case Apply(_, _, _, _) => None
+      case Unary(_, _, _, _) => None
+      case Binary(_, _, _, _, _) => None
+      case IfThenElse(_, _, _, _, _) => None
+      case Let(_, _, _, _, _) => None
+      case LetRec(_, _, _, _, _) => None
+      case Is(_, _, _, _) => None
+      case Tag(_, _, _, _, _) => None
+      case Untag(_, _, _, _, _) => None
+      case Index(_, _, _, _) => None
+      case Tuple(_, _, _) => None
+      case Existential(_, _, _) => None
+      case Universal(_, _, _) => None
+      case NativeConstructor(_, _, _, _) => None
+      case NativeField(_, _, _) => None
+      case NativeMethod(_, _, _, _) => None
+      case UserError(_, _) => None
+      case MatchError(_, _) => None
+      case SwitchError(_, _) => None
+    }
+    AstVisitor.visitExpressionWith(f, exp)
+  }
 
-    // TODO: Implement.
+  /**
+    * Either performs some high-level inlining using `scores`
+    * or returns None to ask traversal to inline inside `exp`
+    */
+  def inlineVisitor(definitions: Map[DefnSym, Constant], scores: Map[DefnSym, Int], exp: Expression)(implicit genSym: GenSym): Option[Expression] =
+  {
+    /**
+      * Returns args and the symbols they should be bound to in required order for substitution and let binding
+      * @param args List[Expression] of arguments
+      * @return (revArgs, argSymbols) where argSymbols is arguments with all arguments replaced with appropriate symbols
+      *         and revArgs are non-var arguments in reverse order
+      */
+    def argsToSym(args: List[Expression]): (List[(Expression,VarSym)], List[VarSym]) =
+    {
+      /**
+        * @param letAcc Arguments and symbols they need to be bound to in reverse order
+        * @param symAcc Symbols to use for inlining in reverse order
+        * @param args Arguments to replace with variables if not variables
+        */
+      @tailrec def reverseArgVars(letAcc: List[(Expression,VarSym)], symAcc: List[VarSym], args: List[Expression]):
```

Comment:
While  tail-recursion is nice, I wonder if this is trying to be tail recursive over a list which will never contain more than 5-15 items? (i.e. the arguments to a function)